File: scripts/analyze_catalog.py
"""
Analyze all 5491 catalog entries from BodyShopThumbnails:
- Distribution of src_grp values
- Cross-reference catalog src_insts vs binary/xml SKIN_ENTRY instances in Skins.package
Uses the working refpack_decomp from parse_catidx.py.
"""
import struct, sys, collections

ISO = 'C:/Users/meyer.n/Downloads/Sims2.iso'
RAR_SIG = b'\x52\x61\x72\x21\x1a\x07\x00'

def r16(b, o): return struct.unpack_from('<H', b, o)[0]
def r32(b, o): return struct.unpack_from('<I', b, o)[0]

def iso_read(f, lba, offset, length):
    f.seek(lba * 2048 + offset)
    return f.read(length)

def refpack_decomp(raw):
    decsz=(raw[6]<<16)|(raw[7]<<8)|raw[8]
    src=raw[9:]; dst=bytearray(decsz); rp=0; wp=0
    while rp<len(src):
        c=src[rp]; lc=0; cc=0; dist=0; last=False
        if c<0x80:
            lc=c&3; cc=((c&0x1C)>>2)+3; dist=((c&0x60)<<3)|src[rp+1]+1; rp+=2
        elif c<0xC0:
            lc=(src[rp+1]&0xC0)>>6; cc=(c&0x3F)+4; dist=((src[rp+1]&0x3F)<<8)|src[rp+2]+1; rp+=3
        elif c<0xE0:
            lc=c&3; cc=((c&0x0C)<<6)|src[rp+3]+5; dist=((c&0x10)<<12)|(src[rp+1]<<8)|src[rp+2]+1; rp+=4
        elif c<0xFC:
            lc=((c&0x1F)<<2)+4; rp+=1
        else:
            lc=c&3; last=True; rp+=1
        for i in range(lc): dst[wp]=src[rp+i]; wp+=1
        rp+=lc
        for i in range(cc):
            if wp>=len(dst): break
            dst[wp]=dst[wp-dist] if wp-dist>=0 else 0; wp+=1
        if last: break
    return bytes(dst[:wp])

with open(ISO, 'rb') as f:
    # Find TSDATA.EXE
    f.seek(16*2048); pvd=f.read(2048)
    root_lba=r32(pvd,158)
    f.seek(root_lba*2048); data=f.read(r32(pvd,166))
    pos=0; tsdata_lba=None
    while pos<len(data):
        rec=data[pos]
        if rec==0: break
        nl=data[pos+32]
        nm=data[pos+33:pos+33+nl].decode('ascii','replace').split(';')[0]
        if nm=='TSDATA.EXE': tsdata_lba=r32(data,pos+2)
        pos+=rec
    print(f"TSDATA.EXE LBA={tsdata_lba}")

    # Find RAR
    rar_off=None
    for base in range(0,64*1024*1024,65529):
        chunk=iso_read(f,tsdata_lba,base,65536)
        i=chunk.find(RAR_SIG)
        if i!=-1: rar_off=base+i; break
    print(f"RAR offset=0x{rar_off:X}")

    # Walk RAR, find BodyShopThumbnails and Skins.package
    pos2=rar_off; bst_doff=None; bst_sz=None; skins_doff=None; skins_sz=None
    while True:
        hdr=iso_read(f,tsdata_lba,pos2,40)
        if len(hdr)<7: break
        bt=hdr[2]; bf=r16(hdr,3); bs=r16(hdr,5)
        if bs==0 or bt==0x7B: break
        if bt==0x74:
            psize=r32(hdr,7); nsize=r16(hdr,26); hoff=32
            if bf&0x0100:
                ex=iso_read(f,tsdata_lba,pos2+32,8)
                psize|=r32(ex,0)<<32; hoff=40
            rnm=iso_read(f,tsdata_lba,pos2+hoff,nsize).decode('ascii','replace').replace('\\','/')
            doff=pos2+bs
            if 'BodyShopThumbnails' in rnm:
                bst_doff=doff; bst_sz=psize
                print(f"Found BodyShopThumbnails at offset=0x{doff:X} size={psize//1024}KB")
            elif 'Skins.package' in rnm and 'Locale' not in rnm:
                skins_doff=doff; skins_sz=psize
                print(f"Found Skins.package at offset=0x{doff:X} size={psize//1024}KB")
            if bst_doff and skins_doff: break
            pos2=doff+psize
        else:
            pos2+=bs+(r32(hdr,7) if bf&0x8000 else 0)

    # Read Skins.package entirely into memory
    print(f"\nReading Skins.package ({skins_sz//1024}KB)...")
    skins_data=iso_read(f,tsdata_lba,skins_doff,skins_sz)
    assert skins_data[:4]==b'DBPF', "Not DBPF!"
    skins_cnt=r32(skins_data,36); skins_ioff=r32(skins_data,40)
    print(f"  {skins_cnt} entries")
    BINARY_SKIN=0xEBCF3E27; XML_SKIN=0x0C1FE246; RKL=0xAC506764
    bin_insts=set(); xml_insts=set(); rkl_insts=set()
    for i in range(skins_cnt):
        p=skins_ioff+i*20
        t=r32(skins_data,p); ins=r32(skins_data,p+8)
        if t==BINARY_SKIN: bin_insts.add(ins)
        elif t==XML_SKIN: xml_insts.add(ins)
        elif t==RKL: rkl_insts.add(ins)
    print(f"  binary SKIN_ENTRY: {len(bin_insts)}, XML: {len(xml_insts)}, RKL: {len(rkl_insts)}")

    # Read BodyShopThumbnails DBPF header to get index location
    bst_hdr=iso_read(f,tsdata_lba,bst_doff,96)
    assert bst_hdr[:4]==b'DBPF'
    bst_cnt=r32(bst_hdr,36); bst_ioff=r32(bst_hdr,40)
    print(f"\nBodyShopThumbnails: {bst_cnt} entries, index at offset {bst_ioff}")

    # Read BST DBPF index
    bst_idx=iso_read(f,tsdata_lba,bst_doff+bst_ioff,bst_cnt*20)
    JPEG=0x856DDBAC; CATALOG_IDX=0x43494745
    cat_off=None; cat_sz=None; jpeg_insts=set()
    for i in range(bst_cnt):
        p=i*20; t=r32(bst_idx,p); ins=r32(bst_idx,p+8); off=r32(bst_idx,p+12); sz=r32(bst_idx,p+16)
        if t==JPEG: jpeg_insts.add(ins)
        elif t==CATALOG_IDX: cat_off=off; cat_sz=sz
    print(f"  {len(jpeg_insts)} JPEG thumbnails, catalog at offset {cat_off} size={cat_sz}")

    # Read and decompress the catalog index
    raw=iso_read(f,tsdata_lba,bst_doff+cat_off,cat_sz)
    dec=refpack_decomp(raw)
    print(f"  Catalog: compressed={cat_sz} -> decompressed={len(dec)}")

    ver=r32(dec,0); cnt=r32(dec,4)
    esz=(len(dec)-8)//cnt if cnt else 0
    print(f"  ver={ver} count={cnt} entry_size={esz}")

    # Analyze all entries
    src_grps=collections.Counter()
    src_insts_set=set()
    matched_bin=0; matched_xml=0; matched_rkl=0; in_bst=0
    seen_grps={}

    for i in range(cnt):
        base=8+i*esz
        if base+esz>len(dec): break
        src_type=r32(dec,base+4)
        src_grp =r32(dec,base+8)
        src_inst=r32(dec,base+12)
        tinst   =r32(dec,base+28)

        src_grps[src_grp]+=1
        src_insts_set.add(src_inst)
        if src_inst in bin_insts: matched_bin+=1
        if src_inst in xml_insts: matched_xml+=1
        if src_inst in rkl_insts: matched_rkl+=1
        if tinst in jpeg_insts:   in_bst+=1

        if src_grp not in seen_grps:
            seen_grps[src_grp]=(i,src_type,src_inst,tinst)

    print(f"\nsrc_grp distribution ({len(src_grps)} unique groups):")
    for grp,c in src_grps.most_common():
        ei,st,si,ti=seen_grps[grp]
        print(f"  0x{grp:08X}: {c:5d} entries  (first at idx {ei}: src_type=0x{st:08X} src_inst=0x{si:08X})")

    print(f"\nCross-reference with Skins.package:")
    print(f"  src_inst in binary SKIN_ENTRY: {matched_bin} / {cnt}")
    print(f"  src_inst in XML SKIN_ENTRY:    {matched_xml} / {cnt}")
    print(f"  src_inst in RKL:               {matched_rkl} / {cnt}")
    print(f"  thumb_inst in BST:             {in_bst} / {cnt}")

    print(f"\nFirst 3 src_insts from catalog:")
    for i in range(3):
        base=8+i*esz
        sg=r32(dec,base+8); si=r32(dec,base+12); ti=r32(dec,base+28)
        print(f"  Entry {i}: src_grp=0x{sg:08X} src_inst=0x{si:08X} thumb_inst=0x{ti:08X}")
        print(f"    -> src_inst in bin_insts: {si in bin_insts}, xml_insts: {si in xml_insts}")

File: scripts/analyzeCatalog.ts
import { closeSync, openSync } from "node:fs";
import { findInstallerLba, findRarOffset, walkRarEntries, type RarEntry } from "./lib/installer.ts";
import { readHeader, readIndex, readResourceBytes } from "./lib/dbpf.ts";
import { decompress } from "./lib/refpack.ts";

// Analyzes every catalog entry from a disc's BodyShopThumbnails package:
//   - the distribution of src_grp values
//   - a cross-reference of catalog src_insts against the binary/XML
//     SKIN_ENTRY instances in Skins.package
//
// Both packages are sealed inside the disc's installer archive, and the
// catalog index itself is RefPack-compressed, so this exercises the same
// installer, DBPF and RefPack machinery the engine's own readers use.

const BINARY_SKIN = 0xebcf3e27;
const XML_SKIN = 0x0c1fe246;
const RESOURCE_KEY_LIST = 0xac506764;
const JPEG_TYPE = 0x856ddbac;
const CATALOG_INDEX_TYPE = 0x43494745;

function hex32(value: number): string {
    return `0x${value.toString(16).padStart(8, "0")}`;
}

function findEntryContaining(
    fd: number,
    lba: number,
    rarOffset: number,
    nameIncludes: readonly string[],
    nameExcludes: readonly string[] = [],
): RarEntry | null {
    let found: RarEntry | null = null;
    walkRarEntries(fd, lba, rarOffset, (entry) => {
        if (
            nameIncludes.some((needle) => entry.name.includes(needle)) &&
            !nameExcludes.some((needle) => entry.name.includes(needle))
        ) {
            found = entry;
            return true;
        }
        return false;
    });
    return found;
}

interface FirstSeen {
    index: number;
    srcType: number;
    srcInstance: number;
    thumbInstance: number;
}

function main(): void {
    const isoPath = process.argv[2];
    if (!isoPath) {
        console.error("usage: analyzeCatalog.ts <path-to-iso>");
        process.exitCode = 1;
        return;
    }

    const fd = openSync(isoPath, "r");
    try {
        const lba = findInstallerLba(fd);
        if (lba === null) {
            console.error("TSDATA.EXE not found");
            process.exitCode = 1;
            return;
        }
        console.log(`TSDATA.EXE LBA=${lba}`);

        const rarOffset = findRarOffset(fd, lba);
        if (rarOffset === null) {
            console.error("RAR archive not found");
            process.exitCode = 1;
            return;
        }
        console.log(`RAR offset=${hex32(rarOffset)}`);

        const bst = findEntryContaining(fd, lba, rarOffset, ["BodyShopThumbnails"]);
        const skins = findEntryContaining(fd, lba, rarOffset, ["Skins.package"], ["Locale"]);
        if (bst === null || skins === null) {
            console.error(`BodyShopThumbnails ${bst ? "found" : "NOT FOUND"}, Skins.package ${skins ? "found" : "NOT FOUND"}`);
            process.exitCode = 1;
            return;
        }
        console.log(`Found BodyShopThumbnails at offset=${hex32(bst.dataOffset)} size=${Math.floor(bst.packedSize / 1024)}KB`);
        console.log(`Found Skins.package at offset=${hex32(skins.dataOffset)} size=${Math.floor(skins.packedSize / 1024)}KB`);

        console.log(`\nReading Skins.package (${Math.floor(skins.packedSize / 1024)}KB)...`);
        const skinsOffset = lba * 2048 + skins.dataOffset;
        const skinsHeader = readHeader(fd, skinsOffset);
        if (skinsHeader === null) {
            console.error("Not DBPF!");
            process.exitCode = 1;
            return;
        }
        const skinsEntries = readIndex(fd, skinsOffset, skinsHeader);
        console.log(`  ${skinsHeader.entryCount} entries`);

        const binInstances = new Set<number>();
        const xmlInstances = new Set<number>();
        const rklInstances = new Set<number>();
        for (const entry of skinsEntries) {
            if (entry.typeIdentifier === BINARY_SKIN) {
                binInstances.add(entry.instanceIdentifier);
            } else if (entry.typeIdentifier === XML_SKIN) {
                xmlInstances.add(entry.instanceIdentifier);
            } else if (entry.typeIdentifier === RESOURCE_KEY_LIST) {
                rklInstances.add(entry.instanceIdentifier);
            }
        }
        console.log(`  binary SKIN_ENTRY: ${binInstances.size}, XML: ${xmlInstances.size}, RKL: ${rklInstances.size}`);

        const bstOffset = lba * 2048 + bst.dataOffset;
        const bstHeader = readHeader(fd, bstOffset);
        if (bstHeader === null) {
            console.error("BodyShopThumbnails is not DBPF!");
            process.exitCode = 1;
            return;
        }
        const bstEntries = readIndex(fd, bstOffset, bstHeader);
        console.log(`\nBodyShopThumbnails: ${bstHeader.entryCount} entries, index at offset ${bstHeader.indexOffset}`);

        const jpegInstances = new Set<number>();
        let catalogEntry: (typeof bstEntries)[number] | null = null;
        for (const entry of bstEntries) {
            if (entry.typeIdentifier === JPEG_TYPE) {
                jpegInstances.add(entry.instanceIdentifier);
            } else if (entry.typeIdentifier === CATALOG_INDEX_TYPE) {
                catalogEntry = entry;
            }
        }
        console.log(`  ${jpegInstances.size} JPEG thumbnails, catalog at offset ${catalogEntry?.offsetInBytes} size=${catalogEntry?.sizeInBytes}`);
        if (catalogEntry === null) {
            console.error("no catalog index found in BodyShopThumbnails");
            process.exitCode = 1;
            return;
        }

        const raw = readResourceBytes(fd, bstOffset, catalogEntry);
        const decompressed = decompress(raw);
        console.log(`  Catalog: compressed=${catalogEntry.sizeInBytes} -> decompressed=${decompressed.length}`);

        const version = decompressed.readUInt32LE(0);
        const count = decompressed.readUInt32LE(4);
        const entrySize = count ? Math.floor((decompressed.length - 8) / count) : 0;
        console.log(`  ver=${version} count=${count} entry_size=${entrySize}`);

        const srcGroupCounts = new Map<number, number>();
        const firstSeenByGroup = new Map<number, FirstSeen>();
        let matchedBinary = 0;
        let matchedXml = 0;
        let matchedRkl = 0;
        let inBst = 0;

        for (let i = 0; i < count; i += 1) {
            const base = 8 + i * entrySize;
            if (base + entrySize > decompressed.length) {
                break;
            }
            const srcType = decompressed.readUInt32LE(base + 4);
            const srcGroup = decompressed.readUInt32LE(base + 8);
            const srcInstance = decompressed.readUInt32LE(base + 12);
            const thumbInstance = decompressed.readUInt32LE(base + 28);

            srcGroupCounts.set(srcGroup, (srcGroupCounts.get(srcGroup) ?? 0) + 1);
            if (binInstances.has(srcInstance)) matchedBinary += 1;
            if (xmlInstances.has(srcInstance)) matchedXml += 1;
            if (rklInstances.has(srcInstance)) matchedRkl += 1;
            if (jpegInstances.has(thumbInstance)) inBst += 1;

            if (!firstSeenByGroup.has(srcGroup)) {
                firstSeenByGroup.set(srcGroup, { index: i, srcType, srcInstance, thumbInstance });
            }
        }

        console.log(`\nsrc_grp distribution (${srcGroupCounts.size} unique groups):`);
        const sortedGroups = [...srcGroupCounts.entries()].sort((a, b) => b[1] - a[1]);
        for (const [group, groupCount] of sortedGroups) {
            const first = firstSeenByGroup.get(group)!;
            console.log(
                `  ${hex32(group)}: ${String(groupCount).padStart(5)} entries  (first at idx ${first.index}: src_type=${hex32(first.srcType)} src_inst=${hex32(first.srcInstance)})`,
            );
        }

        console.log("\nCross-reference with Skins.package:");
        console.log(`  src_inst in binary SKIN_ENTRY: ${matchedBinary} / ${count}`);
        console.log(`  src_inst in XML SKIN_ENTRY:    ${matchedXml} / ${count}`);
        console.log(`  src_inst in RKL:               ${matchedRkl} / ${count}`);
        console.log(`  thumb_inst in BST:             ${inBst} / ${count}`);

        console.log("\nFirst 3 src_insts from catalog:");
        for (let i = 0; i < Math.min(3, count); i += 1) {
            const base = 8 + i * entrySize;
            const srcGroup = decompressed.readUInt32LE(base + 8);
            const srcInstance = decompressed.readUInt32LE(base + 12);
            const thumbInstance = decompressed.readUInt32LE(base + 28);
            console.log(`  Entry ${i}: src_grp=${hex32(srcGroup)} src_inst=${hex32(srcInstance)} thumb_inst=${hex32(thumbInstance)}`);
            console.log(`    -> src_inst in bin_insts: ${binInstances.has(srcInstance)}, xml_insts: ${xmlInstances.has(srcInstance)}`);
        }
    } finally {
        closeSync(fd);
    }
}

main();
